internal/dlog/calc.go

```diff
@@ -69,6 +69,45 @@ func (c *CalcZp) WithBound(bound *big.Int) *CalcZp {
 	return c
 }
 
+func (c *CalcZp) RunBabyStepGiantStep(h, g *big.Int, retChan chan *big.Int, errChan chan error) {
+	ret, err := c.BabyStepGiantStep(h, g)
+	retChan <- ret
+	errChan <- err
+}
+
+// BabyStepGiantStepWithNeg uses the baby-step giant-step method to
+// compute the discrete logarithm in the Zp group if the answer is
+// within [-bound, bound]. It does so by running two goroutines, one
+// for negative answers and one for positive.
+func (c *CalcZp) BabyStepGiantStepWithNeg(h, g *big.Int) (*big.Int, error) {
```

Comment:
`CalcZp` struct was designed to keep configuration of the dlog calculator in order to avoid having too many (possibly optional) parameters for the baby step giant step methods. So, for unbounded search, we use `calc.BabyStepGiantStep(...)`, and for bounded search we do `calc.WithBound(b).BabyStepGiantStep(...)`, rather than having an optional `bound` parameter for the `BabyStepGiantStep` method. I think whether to search for the answer also among negative solutions is another "configuration option" of the dlog calculator. So my suggestion is to add a `bool` field `neg` to `CalcZp` struct (which is `false` by default), then create a function `WithNeg()` that returns a new instance of `CalcZp` with the `neg` field set to `true`.

````go
func (c *CalcZp) WithNeg() *CalcZp {
	return &CalcZp{
		bound: c.bound,
		m:     c.m,
		p:     c.p,
		neg:   true,
	}
}
````

In your implementation, we have exported methods: `BabyStepGiantStep`, `BabyStepGiantStepWithNeg` and `RunBabyStepGiantStep`. I would prefer exporting a single method (`BabyStepGiantStep`) that would, based on configuration of `CalcZp` (if `neg` is `true`, which would be set by the caller via `WithNeg()` method) either do a dlog search among positive numbers, or among positive + negative numbers. This is also the method that should take care of the communication between goroutines. 

I also think we're missing a test for the concurrent version of the algorithm.

Reply:
I don't see how to avoid having at least a helping function. Is it acceptable for you to have a private function say `runBabyStepGianStep` that will be called from `BabyStepGiantStep`?

Reply:
Sure